Improve error reporting of named argument mismatch

radeusgd opened the pull request:
### Pull Request Description

- Related to further work discussed in #7359
- Closes #12316

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

radeusgd committed:
Separate error for named argument mismatch

radeusgd committed:
CR from previous PR - use autoscoping

radeusgd committed:
include `call_target` in No_Such_Argument, like Not_Invokable

radeusgd committed:
refactor Widgets hack to be more precise catching the error

radeusgd committed:
intercept No_Such_Argument and provide a better error for Delimited argument rename

radeusgd committed:
add tests

radeusgd committed:
reorder

radeusgd committed:
javafmt

radeusgd committed:
also handle the indirect case

JaroslavTulach commented on test/micro-distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:
```diff
@@ -40,6 +40,12 @@ type Not_Invokable
+type No_Such_Argument
+    Error argument_name call_target
+
+    to_display_text self = "The named argument `"+self.argument_name.to_text+"` did not match any argument names. Perhaps it is misspelled?"
```
I see a [CI failure](https://github.com/enso-org/enso/actions/runs/13163967243/job/36739521558?pr=12238#step:8:5096):
```
org.enso.interpreter.test.builtins.BuiltinsInvocationTest.invokeBuiltinWithWrongArguments_ShouldNotCrash failed: java.lang.AssertionError: Should have Type error as payload
Expected: a string containing "Type error"
but: was "The named argument `self` did not match any argument names. Perhaps it is misspelled?", took 1.44 sec
at org.enso.interpreter.test.builtins.BuiltinsInvocationTest.invokeBuiltinWithWrongArguments_ShouldNotCrash(BuiltinsInvocationTest.java:41)
at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
at java.lang.reflect.Method.invoke(Method.java:580)
``` 
shouldn't we continue to throw `.makeNotInvokable(callable);` and just add more info to it? Otherwise we are dealing with incompatible change.

radeusgd replied:
Good point. I was thinking that it's better to split, but I think you convinced me that it's better to keep the change compatible.

I think I will keep the `No_Such_Argument` builtin type but will not throw it directly, instead I will add it as a 'cause' to `Not_Invokable`. This way we can get a nicer message and check for the no-such-argument case when we want to, but the error type thrown is unchanged.

radeusgd replied:
API change reverted in 09889dad27a26a27857a36c519a30324f8eaa086, now it contains a `cause` that can be `No_Such_Argument` instance.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/CurryNode.java:
```diff
@@ -78,7 +78,8 @@ private void initializeOversaturatedCallNode(
               defaultsExecutionMode,
-              argumentsExecutionMode);
+              argumentsExecutionMode,
+              true);
```
Why do we need new `boolean` argument? Isn't `postApplicationSchema.getOversaturatedArguments()` good enough indicator that there are oversaturated arguments?

radeusgd replied:
But the `create` method/constructor does not see `postApplicationSchema`, it sees the result of `postApplicationSchema.getOversaturatedArguments()` which is just a `CallArgumentInfo[] schema`.

From inside of `InvokeCallableNodeGen.create` you can no longer tell if the arguments being passed were 'oversaturated' or 'normal' - you just get some schema, that is the same kind as in all different places where `InvokeCallableNode` is constructed.

That's why this argument is added to be able to distinguish the situations at call-site.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java:
```diff
@@ -104,11 +104,16 @@ public boolean shouldExecute() {
   private final int thatArgumentPosition;
 
   private final ArgumentsExecutionMode argumentsExecutionMode;
+  private final CallArgumentInfo[] schema;
```
Assuming schema elements are also supposed to be final, one must add `@CompilationFinal(dimentions=1)`

radeusgd replied:
Thanks, added it.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java:
```diff
@@ -375,8 +381,17 @@ static Object doPolyglot(
       Object callable, VirtualFrame callerFrame, State state, Object[] arguments) {
-    Atom error = EnsoContext.get(this).getBuiltins().error().makeNotInvokable(callable);
-    throw new PanicException(error, this);
+    if (isForOversaturatedArguments && schema.length >= 1 && schema[0].isNamed()) {
```
Is this a compilation constant? Try `CompilerAsserts.partialEvaluationContant` or co.

radeusgd replied:
Tried in e645b0571576ef5b520287248c4b38537be3183d, is that how it is supposed to be used?

JaroslavTulach commented on distribution/lib/Standard/Visualization/0.0.0-dev/src/Widgets.enso:
```diff
@@ -27,9 +27,12 @@ get_widget_json value call_name argument_names uuids="{}" =
-        return_target err = err.payload.target
-        Panic.catch Not_Invokable handler=return_target
-            annotation value cache=cache
+        Panic.catch No_Such_Argument (annotation value cache=cache) caught_panic->
```
This code change shouldn't be necessary if the API change was compatible.

radeusgd replied:
We wanted to change the code anyway - previously the `catch` was too broad - it was catching _any_ `Not_Invokable` error, not necessarily the correct one.

But yes I can remove the new type and add extra info to the current error. The code change will stay but will be less different.

JaroslavTulach reviewed (commented):
- adventurous: I am afraid to touch `CurryNode` & co. myself
- changing the type of thrown panic is an **incompatible change**
    - better to keep the atom and add an `info` field to it
- are the benchmarks result OK? 
    - Probably they are (as these are just exceptional paths and they are unlikely tested).
- one minor `@CompilationFinal` note - otherwise looks decent

radeusgd committed:
CR: revert to Not_Invokable error for compatibility, but add cause information

radeusgd committed:
after CR: adapt my changes to the changed API

radeusgd committed:
CR: CompilationFinal array

radeusgd committed:
CR: assert compilation constant

radeusgd committed:
shorter error message

radeusgd committed:
refactor and document annotation invocation logic

radeusgd committed:
fix changelog

radeusgd committed:
add to changelog

radeusgd commented:
Double checked in the GUI:
![image](https://github.com/user-attachments/assets/5baf63e3-151b-4f86-b210-7c744e7b3f72)

The widgets that rely on cache seem to all work fine - so the tweaking of the annotation invocation code seems to be all good.

radeusgd commented:
> are the benchmarks result OK?
> * Probably they are (as these are just exceptional paths and they are unlikely tested).

I will schedule a run once I ensure that the tests are passing.

radeusgd committed:
update message to conform with test

radeusgd commented:
![image](https://github.com/user-attachments/assets/e52694fb-cbf6-4165-965d-eedef9b8357d)

radeusgd committed:
javafmt

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:
```diff
@@ -310,12 +310,32 @@ type Not_Invokable
+
+       Arguments:
+       - argument_name: The name of the argument that was not found.
+    Error argument_name
```
Remove `## PRIVATE` use:
```suggestion
    private Error argument_name
```
possibly remove the documentation comment altogether.

radeusgd replied:
This is against the convention that we are using for all the errors.

I think changes in this PR should be consistent with the rest of the codebase - if we want to change the convention let's discuss and consider rewriting the whole `Common.enso` file.

radeusgd replied:
We don't want the constructor to be `private`, this breaks the very simple check I've implemented like `payload.cause == (No_Such_Argument.Error "foo")`, at least outside of Base.

The `PRIVATE` is to avoid cluttering Component Browser only, I guess at some point we may remove that tag from errors.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:
```diff
@@ -310,12 +310,32 @@ type Not_Invokable
        - target: The called object.
-    Error target
+       - cause: Additional information about what may have caused the error.
+    Error target cause
```
This is still an incompatible change for the creators of this object. To make the change compatible use:
```suggestion
    Error target cause=Nothing
```

radeusgd replied:
Ok

radeusgd replied:
I've tried and unfortunately constructors of builtins do not like default arguments.

With the change above, the code
```
from Standard.Base import all
import Standard.Base.Errors.Common.Not_Invokable

main =
    without_cause = Not_Invokable.Error "thing"
    IO.println without_cause
```
still prints
```
Not_Invokable.Error target='thing' cause=_
```

I think the defaultness has to be handled in engine and we don't have capability to do that.

Technically, even with the default argument this change would still breaking as any pattern matching of `Not_Invokable.Error _` also will now mismatch the amount of fields.

We could try to have 2 constructors for `Not_Invokable` - the old one and the new one. But I think this is complicating stuff unnecessarily. Also not sure if the builtins machinery would currently allow this, as it would no longer be a `UniquelyConstructibleBuiltin`.

JaroslavTulach commented on distribution/lib/Standard/Visualization/0.0.0-dev/src/Widgets.enso:
```diff
@@ -27,9 +28,21 @@ get_widget_json value call_name argument_names uuids="{}" =
+                    ## If the cache argument was not expected, we ignore the problem
+                       and return just the result of `f value` - held by
+                       the target of the Not_Invokable error.
+                    was_not_expecting_cache_argument = caught_panic.payload.cause == (No_Such_Argument.Error "cache")
```
Question to myself: Why do I find this unreadable? Would following be better?

```ruby
f : Function ->
    Panic.catch Not Invocable (f value cache=cache) caught_panic -> case caught.panic.payload.cause of
       No_Such_Argument.Error "cache" -> caught_panic.payload.target
       _ -> Panic.throw caught_panic
```

Maybe, but `caught.panic.payload.cause` is still quite long...

JaroslavTulach commented on distribution/lib/Standard/Visualization/0.0.0-dev/src/Widgets.enso:
```diff
@@ -27,9 +28,21 @@ get_widget_json value call_name argument_names uuids="{}" =
+            f : Function ->
+                # We pass the self value and the cache to the function.
+                Panic.catch Not_Invokable (f value cache=cache) caught_panic->
+                    ## If the cache argument was not expected, we ignore the problem
```
Why do you use `##` instead of `#` here?
```suggestion
                    # If the cache argument was not expected, we ignore the problem
```

radeusgd replied:
Because the `##` allows to write a multi-line comment, and I don't want to create lines 300+ characters long.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java:
```diff
@@ -375,7 +385,16 @@ static Object doPolyglot(
   public Object invokeGeneric(
       Object callable, VirtualFrame callerFrame, State state, Object[] arguments) {
-    Atom error = EnsoContext.get(this).getBuiltins().error().makeNotInvokable(callable);
+    Atom cause = null;
```
The same code is already present in `IndirectInvokeCallableNode` once. Can we refactor it to a single method?

radeusgd replied:
Hmmm, perhaps a good point.

Any suggestions where this common function could live? Put it as `static` in `InvokeCallableNode` and call from `IndirectInvokeCallableNode`?

radeusgd replied:
Done in [97b7d27](https://github.com/enso-org/enso/pull/12238/commits/97b7d274adecd13f52f64b41fde44b54ccd0c1de)

radeusgd committed:
CR: make cause default to Nothing

radeusgd committed:
CR: extract common logic

radeusgd committed:
Revert "CR: make cause default to Nothing"

This reverts commit 38ca032532ccf3a8eebd9ddd9174a3096fe04f3f.

radeusgd commented:
Running benchmarks:
- https://github.com/enso-org/enso/actions/runs/13198413475
  - followup after merging `develop` to verify some 'unsure' results: https://github.com/enso-org/enso/actions/runs/13374457813
- https://github.com/enso-org/enso/actions/runs/13197153956

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/improve-mismatched-named-argument-error

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/improve-mismatched-named-argument-error

radeusgd committed:
add a currying test

radeusgd committed:
make sure API generator uses UNIX line endings - otherwise running it on Windows caused tons of spurious differences (CRLF vs LF)

radeusgd committed:
updating Base API:
- added `No_Such_Argument` type
- Not_Invokable type got one more field (`cause`)

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/improve-mismatched-named-argument-error

radeusgd commented:
The suspicious benchmark seems to be back to normal on new runs, it looks like it was some variation within the variance of the benchmark.
![image](https://github.com/user-attachments/assets/a93115a6-3237-4ee8-924f-ad3e5b33a855)

radeusgd committed:
normalize checked out line endings in signature test
